# Bayes' Theorem

## Why Do We Need Bayes' Theorem?

In many real-world situations, we observe some evidence and want to update our belief about an event.

Bayes' Theorem answers the question:

> Given that event $B$ has occurred, what is the probability that event $A$ actually occurred?

Mathematically, we want:

$$
P(A|B)
$$

This quantity is called the **Posterior Probability**.

---

## Example Intuition

Suppose:

- A medical test result is positive.
- We want to know whether the patient actually has the disease.

The test result is the **evidence**.

The disease status is what we want to infer.

Bayes' Theorem allows us to reverse the probability.

Instead of:

$$
P(\text{Positive Test} \mid \text{Disease})
$$

we want:

$$
P(\text{Disease} \mid \text{Positive Test})
$$

These two probabilities are generally **not equal**.

---

# Derivation of Bayes' Theorem

Bayes' Theorem is not a new formula.

It is obtained directly from the definition of conditional probability.

## Step 1: Conditional Probability of A Given B

From the definition of conditional probability,

$$
P(A|B)
=
\frac{P(A \cap B)}{P(B)}
\qquad P(B) \neq 0
$$

This means:

> Among all outcomes where $B$ occurred, how many also belong to $A$?

---

## Step 2: Conditional Probability of B Given A

Similarly,

$$
P(B|A)
=
\frac{P(A \cap B)}{P(A)}
\qquad P(A) \neq 0
$$

Rearranging,

$$
P(A \cap B)
=
P(B|A)P(A)
$$

This gives another way to express the joint probability.

---

## Step 3: Substitute into Conditional Probability Formula

From Step 1,

$$
P(A|B)
=
\frac{P(A \cap B)}{P(B)}
$$

Substituting

$$
P(A \cap B)=P(B|A)P(A)
$$

gives

$$
P(A|B)
=
\frac{P(B|A)P(A)}{P(B)}
$$

---

# Bayes' Theorem

$$
P(A|B)
=
\frac{P(B|A)P(A)}{P(B)}
$$

where

$$
P(B) \neq 0
$$

---

# Understanding Every Component

Bayes' Theorem can be written as

$$
P(A|B)
=
\frac{P(B|A)P(A)}{P(B)}
$$

Each term has a specific meaning.

---

## 1. Posterior Probability

$$
P(A|B)
$$

Probability of $A$ after observing $B$.

This is the quantity we want to compute.

**Think:** Updated Belief

---

## 2. Likelihood

$$
P(B|A)
$$

Probability of observing evidence $B$ assuming $A$ is true.

**Think:** How likely is the evidence if my hypothesis is correct?

---

## 3. Prior Probability

$$
P(A)
$$

Probability of $A$ before seeing any evidence.

**Think:** Initial Belief

---

## 4. Evidence (Marginal Probability)

$$
P(B)
$$

Overall probability of observing $B$.

**Think:** Normalization Factor

---

# Memory Formula

$$
\text{Posterior}
=
\frac{\text{Likelihood} \times \text{Prior}}{\text{Evidence}}
$$

Or

```text
Updated Belief
=
Evidence × Initial Belief
-------------------------
Normalization
```

---

# Medical Diagnosis Example

Define:

- $C$ = Person has disease
- $NC$ = Person does not have disease
- $+$ = Test result is positive

Suppose

$$
P(C)=0.01
$$

Only 1% of people have the disease.

Therefore,

$$
P(NC)=0.99
$$

---

## Test Characteristics

### True Positive Rate

$$
P(+|C)=0.9
$$

If a person has the disease, the test correctly detects it 90% of the time.

### False Positive Rate

$$
P(+|NC)=0.1
$$

If a person does not have the disease, the test still reports positive 10% of the time.

---

# Goal

We want to compute

$$
P(C|+)
$$

which means:

> Given that the test result is positive, what is the probability that the person actually has the disease?

---

# Probability Tree

```text
Population
│
├── Disease (1%)
│    ├── Positive (90%)
│    └── Negative (10%)
│
└── No Disease (99%)
     ├── Positive (10%)
     └── Negative (90%)
```

---

# Total Probability Theorem

Before applying Bayes' Theorem, we need

$$
P(+)
$$

the overall probability of obtaining a positive test result.

A positive result can occur in two ways:

1. Diseased person tests positive
2. Healthy person tests positive

Therefore,

$$
P(+)
=
P(+ \cap C)
+
P(+ \cap NC)
$$

Using joint probability,

$$
P(+)
=
P(+|C)P(C)
+
P(+|NC)P(NC)
$$

This is known as the **Total Probability Theorem**.

---

## Calculation

$$
P(+)
=
0.9(0.01)
+
0.1(0.99)
$$

$$
=
0.009+0.099
$$

$$
=
0.108
$$

Thus,

$$
P(+)=0.108
$$

---

# Applying Bayes' Theorem

Using

$$
P(C|+)
=
\frac{P(+|C)P(C)}{P(+)}
$$

Substituting values,

$$
P(C|+)
=
\frac{0.9\times0.01}{0.108}
$$

$$
=
\frac{0.009}{0.108}
$$

$$
=
0.0833
$$

---

# Final Result

$$
P(C|+)
=
8.33\%
$$

---

# Important Interpretation

Many of us  expect the answer to be 90%.


Why?

Because they confuse

$$
P(+|C)
$$

with

$$
P(C|+)
$$

But these are completely different probabilities.

We know

$$
P(+|C)=90\%
$$

Yet

$$
P(C|+)=8.33\%
$$

A positive test does **not** imply a 90% chance of disease.

The disease is very rare, so false positives dominate.

This phenomenon is called the **Base Rate Effect**.

---

# Relationship Between Key Probability Concepts

## Conditional Probability

Answers:

> Given that $B$ occurred, what is the probability of $A$?

$$
P(A|B)
=
\frac{P(A \cap B)}{P(B)}
$$

---

## Joint Probability

Answers:

> What is the probability that both $A$ and $B$ occur together?

$$
P(A \cap B)
=
P(A|B)P(B)
$$

or

$$
P(A \cap B)
=
P(B|A)P(A)
$$

---

## Total Probability

Answers:

> What is the overall probability of event $B$?

$$
P(B)
=
P(B|A)P(A)
+
P(B|A^c)P(A^c)
$$

More generally,

$$
P(B)
=
\sum_i P(B|A_i)P(A_i)
$$

where $\{A_i\}$ forms a partition of the sample space.

---

## Bayes' Theorem

Answers:

> Given evidence $B$, how should we update our belief about $A$?

$$
P(A|B)
=
\frac{P(B|A)P(A)}{P(B)}
$$

---

# Machine Learning Connection

Bayes' Theorem is the mathematical foundation of:

- Naive Bayes Classifier
- Bayesian Machine Learning
- Bayesian Networks
- Spam Detection
- Fraud Detection
- Medical Diagnosis Systems
- Information Retrieval
- Recommendation Systems

In ML terminology:

| Bayes Term | ML Interpretation |
|------------|------------------|
| $P(A)$ | Prior belief about a class |
| $P(B|A)$ | Likelihood of observing features |
| $P(B)$ | Evidence |
| $P(A|B)$ | Posterior probability of class |

For classification,

$$
P(\text{Class}|\text{Features})
=
\frac{P(\text{Features}|\text{Class})P(\text{Class})}{P(\text{Features})}
$$

This is exactly what the **Naive Bayes Classifier** computes.

---

# Common Interview Question

## Is Bayes' Theorem a New Formula?

**No.**

Bayes' Theorem is obtained by combining:

### Conditional Probability

$$
P(A|B)
=
\frac{P(A\cap B)}{P(B)}
$$

### Joint Probability

$$
P(A\cap B)
=
P(B|A)P(A)
$$

and substituting one into the other.

---

# Common Mistake (Very Important)

we often think

$$
P(A|B)=P(B|A)
$$

This is almost never true.

For example,

$$
P(+|C)=90\%
$$

but

$$
P(C|+)=8.33\%
$$

Bayes' Theorem exists precisely because these two probabilities are different.

---

# Quick Revision Box

### Conditional Probability

$$
P(A|B)
=
\frac{P(A\cap B)}{P(B)}
$$

Think:

> Given B

---

### Joint Probability

$$
P(A\cap B)
=
P(B|A)P(A)
$$

Think:

> A AND B

---

### Total Probability

$$
P(B)
=
P(B|A)P(A)
+
P(B|A^c)P(A^c)
$$

Think:

> All possible ways B can happen

---

### Bayes' Theorem

$$
P(A|B)
=
\frac{P(B|A)P(A)}{P(B)}
$$

Think:

> Update belief after seeing evidence

---

# One-Line Summary

> Bayes' Theorem updates our prior belief using new evidence to obtain a posterior belief.

```text
Conditional Probability
        ↓
Joint Probability
        ↓
Total Probability
        ↓
Bayes' Theorem
        ↓
Naive Bayes & Bayesian ML
```

# Bayes' Theorem

## Intuition

Bayes' Theorem helps us answer questions of the form:

> Given that some evidence has occurred, what is the probability that a particular event is actually true?

In simple words:

> Bayes updates our belief after observing new information.

---

# The Core Question

Normally we know:

$$
P(B|A)
$$

Probability of observing B when A is true.

But often we want:

$$
P(A|B)
$$

Probability that A is true after observing B.

Bayes Theorem helps us reverse the probability.

---

# Bayes' Theorem Formula

Let:

- A = Event of interest
- B = Observed evidence

Then

$$
P(A|B)
=
\frac{P(B|A)P(A)}
{P(B)}
$$

---

# Meaning of Each Term

## Posterior Probability

What we want to find:

$$
P(A|B)
$$

Probability that A is true after observing B.

---

## Likelihood

How likely we are to observe B if A is true:

$$
P(B|A)
$$

---

## Prior Probability

Initial belief before seeing evidence:

$$
P(A)
$$

---

## Evidence

Overall probability of observing B:

$$
P(B)
$$

---

# One-Line Summary

Bayes Theorem tells us:

> Posterior = Likelihood × Prior / Evidence

---

# When Do We Use Bayes Theorem?

Use Bayes whenever the question contains:

- Given that...
- If we observed...
- If a test is positive...
- If an email contains a word...
- If a customer clicked an ad...

and asks:

> What is the probability of the underlying event?

---

# Common Applications

## 1. Spam Detection

Question:

> If an email contains the word "offer", what is the probability it is spam?

---

## 2. Medical Diagnosis

Question:

> If a patient tests positive, what is the probability they actually have the disease?

---

## 3. Fraud Detection

Question:

> Given a suspicious transaction, what is the probability it is fraud?

---

## 4. Machine Learning

Naive Bayes classifier uses Bayes Theorem to classify:

- Spam / Not Spam
- Positive / Negative Review
- Topic Classification

---


# Example 1: Spam Detection

## Problem Statement

Suppose we want to classify emails as:

- **Spam**
- **Not Spam (Desired Email)**

Let:

- $S$ = Email is Spam
- $O$ = Email contains the word **"offer"**

We are given:

$$
P(S) = 0.3
$$

Meaning:

> 30% of all emails are spam.

---

## Likelihood

80% of spam emails contain the word **"offer"**.

$$
P(O \mid S) = 0.8
$$

Meaning:

> If an email is spam, there is an 80% chance it contains the word "offer".

---

## Desired (Non-Spam) Emails

Since 30% are spam:

$$
P(\bar{S}) = 1 - P(S)
$$

$$
P(\bar{S}) = 1 - 0.3
$$

$$
P(\bar{S}) = 0.7
$$

Meaning:

> 70% of emails are normal (non-spam).

---

Only 10% of normal emails contain the word **"offer"**.

$$
P(O \mid \bar{S}) = 0.1
$$

Meaning:

> If an email is not spam, there is only a 10% chance it contains the word "offer".

---

# Visualizing with 100 Emails

Assume we have **100 emails**.

### Spam Emails

$$
100 \times 0.3 = 30
$$

So:

- Spam Emails = 30

### Normal Emails

$$
100 \times 0.7 = 70
$$

So:

- Normal Emails = 70

---

## Emails Containing "Offer"

### Among Spam Emails

80% contain "offer":

$$
30 \times 0.8 = 24
$$

Therefore:

- Spam + Offer = 24

---

### Among Normal Emails

10% contain "offer":

$$
70 \times 0.1 = 7
$$

Therefore:

- Normal + Offer = 7

---

## Total Emails Containing "Offer"

$$
24 + 7 = 31
$$

Thus:

$$
P(O) = \frac{31}{100}
$$

$$
P(O) = 0.31
$$

---

# Applying Bayes' Theorem

We want:

$$
P(S \mid O)
$$

Meaning:

> What is the probability that an email is spam given that it contains the word "offer"?

Bayes' Theorem:

$$
P(S \mid O)
=
\frac{P(O \mid S)P(S)}
{P(O)}
$$

Substituting values:

$$
P(S \mid O)
=
\frac{0.8 \times 0.3}
{0.31}
$$

$$
=
\frac{0.24}{0.31}
$$

$$
=
0.774
$$

---

# Final Answer

$$
P(S \mid O) = 0.774
$$

or

$$
77.4\%
$$

---

# Interpretation

Even though only **30% of all emails are spam**, once we observe that the email contains the word **"offer"**, the probability that it is spam increases to:

$$
77.4\%
$$

This is exactly what Bayes' Theorem does:

> It updates our belief after observing new evidence.

# Example 2: Medical Diagnosis Using Bayes' Theorem

## Problem Statement

Suppose:

- A disease affects 60% of the population.
- A medical test is 99% accurate.

A patient tests positive.

We want to find:

> What is the probability that the patient actually has the disease?

---

# Define Events

Let:

- $D$ = Person has the disease
- $+$ = Test result is positive

We need to find:

$$
P(D|+)
$$

Meaning:

> Probability that the person has the disease given that the test result is positive.

---

# Given Information

## Prior Probability

60% of people have the disease.

$$
P(D)=0.6
$$

Therefore:

$$
P(\bar D)=0.4
$$

Meaning:

> 40% of people do not have the disease.

---

## Test Accuracy

The test correctly identifies infected people 99% of the time.

$$
P(+|D)=0.99
$$

This is called the **True Positive Rate**.

---

The test incorrectly reports positive for healthy people 1% of the time.

$$
P(+|\bar D)=0.01
$$

This is called the **False Positive Rate**.

---

# Visualizing with 100 People

Assume we have **100 people**.

---

## People With Disease

$$
100 \times 0.6 = 60
$$

So:

- Diseased = 60

---

## People Without Disease

$$
100 \times 0.4 = 40
$$

So:

- Healthy = 40

---

# Apply Test Accuracy

## Positive Results Among Diseased People

99% of diseased people test positive.

$$
60 \times 0.99
=
59.4
$$

So:

- Disease + Positive Test = 59.4

---

## Positive Results Among Healthy People

1% of healthy people falsely test positive.

$$
40 \times 0.01
=
0.4
$$

So:

- Healthy + Positive Test = 0.4

---

# Total Positive Test Results

Positive results come from:

1. Truly diseased people
2. False positives

Therefore:

$$
59.4 + 0.4
=
59.8
$$

Thus:

$$
P(+)=\frac{59.8}{100}
$$

$$
P(+)=0.598
$$

---

# Applying Bayes' Theorem

We want:

$$
P(D|+)
$$

Bayes' Theorem:

$$
P(D|+)
=
\frac{P(+|D)P(D)}
{P(+)}
$$

Substituting the values:

$$
P(D|+)
=
\frac{0.99 \times 0.6}
{0.598}
$$

$$
=
\frac{0.594}{0.598}
$$

$$
=
0.9933
$$

---

# Final Answer

$$
P(D|+)
=
0.9933
$$

or

$$
99.33\%
$$

---

# Interpretation

Before taking the test:

$$
P(D)=60\%
$$

After observing a positive test result:

$$
P(D|+)=99.33\%
$$

The positive test significantly increases our confidence that the person actually has the disease.

---

# Why Bayes' Theorem Is Needed

The test accuracy tells us:

$$
P(+|D)
$$

Probability of getting a positive result **if the person already has the disease**.

But the question asks:

$$
P(D|+)
$$

Probability that the person has the disease **after observing a positive result**.

These two probabilities are **not the same**.

Bayes' Theorem allows us to reverse the conditional probability.

---

# Key Learning

Bayes' Theorem answers questions of the form:

> Given some observed evidence, what is the probability that the underlying event is actually true?

In this example:

- Event = Person has disease
- Evidence = Positive test result

Bayes updates our belief after seeing the evidence.

---

# One-Line Summary

A positive test result does not directly tell us the probability of having a disease. Bayes' Theorem combines the disease prevalence and test accuracy to compute the actual probability that the patient has the disease.

# Population and Sample

## Why Do We Need Sampling?

In real-world problems, collecting data from every individual is often impossible.

Examples:

- Estimating the average salary of all Amazon employees.
- Estimating the average height of all Indians.
- Estimating the average spending of all customers on Amazon.

The complete collection of all observations is called the **Population**.

---

# Population

A **Population** is the entire set of observations that we are interested in studying.

Examples:

- All Amazon employees.
- All Walmart employees.
- All customers of a company.
- All voters in a country.

Mathematically,

$$
N = \text{Population Size}
$$

where:

- $N$ denotes the number of observations in the population.
- Usually very large.

---

## Example

Suppose we want to know the average salary of all Amazon employees.

Population:

```text
Amazon Employees
      ↓
600 Million Employees
```

Here,

$$
N = 600,000,000
$$

Studying every employee is expensive and time-consuming.

---

# Why Not Use the Entire Population?

Because:

1. Cost is very high.
2. Data collection is slow.
3. Some populations are practically infinite.
4. Data may constantly change.

Therefore, we use a **Sample**.

---

# Sample

A **Sample** is a subset of the population selected for analysis.

The sample should ideally be selected randomly.

Mathematically,

$$
n = \text{Sample Size}
$$

where

$$
n < N
$$

---

## Example

Population:

$$
N = 600,000,000
$$

Choose

$$
n = 2000
$$

employees randomly.

Then,

```text
Population
   ↓
Random Sampling
   ↓
2000 Employees
```

The 2000 employees form our sample.

---

# Random Sampling

A good sample should be selected randomly.

Every observation should have an equal chance of being selected.

Why?

Because random samples reduce bias and better represent the population.

---

## Visual Idea

```text
Population
-----------------------------------------
| 0 | 1 | 2 | 3 | 4 | ..... | 99 |
-----------------------------------------

Randomly choose:

3, 18, 27, 54, 89

These become the sample.
```

In Python:

```python
import numpy as np

np.random.randint(0, 100, 5)
```

Possible output:

```python
array([3, 18, 27, 54, 89])
```

This selects 5 random observations from the population.

---

# Sample Size

The number of observations selected is called the sample size.

Example:

$$
n = 5
$$

means

```text
5 random observations
```

are selected from the population.

---

# Population Mean

The true average of the entire population is called the Population Mean.

It is denoted by:

$$
\mu
$$

(read as "mu")

Formula:

$$
\mu
=
\frac{\sum_{i=1}^{N} x_i}{N}
$$

where:

- $N$ = population size
- $x_i$ = population observations

---

# Sample Mean

The average calculated using only sample observations is called the Sample Mean.

It is denoted by:

$$
\bar{x}
$$

(read as "x bar")

Formula:

$$
\bar{x}
=
\frac{\sum_{i=1}^{n} x_i}{n}
$$

where:

- $n$ = sample size
- $x_i$ = sample observations

---

# Main Goal of Statistics

The central goal of statistics is:

> Use information from a sample to estimate properties of the population.

In particular:

$$
\bar{x}
\rightarrow
\mu
$$

Sample Mean estimates Population Mean.

---

# Law of Large Numbers (Intuition)

As sample size increases,

$$
n \rightarrow \infty
$$

the sample mean approaches the population mean.

Mathematically,

$$
\bar{x}
\rightarrow
\mu
$$

This idea is known as the **Law of Large Numbers**.

---

# Important Intuition

Small Sample:

```text
n = 5
```

May not represent the population well.

Large Sample:

```text
n = 5000
```

Usually gives a much better estimate of the population.

Therefore,

$$
n \uparrow
\quad \Rightarrow \quad
\bar{x} \text{ becomes more reliable}
$$

---

# Statistical Inference

The entire field of statistics is built around the idea:

```text
Population
     ↓
Random Sampling
     ↓
Sample
     ↓
Sample Statistics
     ↓
Estimate Population Parameters
```

---

# Terminology Summary

| Concept | Symbol | Meaning |
|----------|----------|----------|
| Population Size | $N$ | Total observations |
| Sample Size | $n$ | Selected observations |
| Population Mean | $\mu$ | True average |
| Sample Mean | $\bar{x}$ | Estimated average |
| Population | Entire dataset | All observations |
| Sample | Subset | Randomly selected observations |

---

# Machine Learning Connection

Almost all Machine Learning models are trained using samples, not populations.

Examples:

- Training Dataset → Sample
- Entire Real World Data → Population

ML assumes:

```text
Training Data
      ↓
Represents
      ↓
Real Population
```

A model learns from the sample and generalizes to the population.

This is why proper sampling is one of the most important concepts in Machine Learning.

---

# One-Line Summary

> A sample is a randomly selected subset of a population, and statistics uses sample information to estimate population properties.

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt



In [20]:
id = "1PLWv2uhMSFlyklKRhxQAyONNwW-Fho0R"
print("https://drive.google.com/uc?export=download&id="+id)


https://drive.google.com/uc?export=download&id=1PLWv2uhMSFlyklKRhxQAyONNwW-Fho0R


In [ ]:
!wget "https://drive.google.com/uc?export=download&id=1PLWv2uhMSFlyklKRhxQAyONNwW-Fho0R" -O marketing_data.csv
#! is to run on terminal
#wget is for web get

--2026-06-14 06:50:05--  https://drive.google.com/uc?export=download&id=1PLWv2uhMSFlyklKRhxQAyONNwW-Fho0R
Resolving drive.google.com (drive.google.com)... 2404:6800:4000:101d::8b, 2404:6800:4000:101d::71, 2404:6800:4000:101d::65, ...
Connecting to drive.google.com (drive.google.com)|2404:6800:4000:101d::8b|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://drive.usercontent.google.com/download?id=1PLWv2uhMSFlyklKRhxQAyONNwW-Fho0R&export=download [following]
--2026-06-14 06:50:06--  https://drive.usercontent.google.com/download?id=1PLWv2uhMSFlyklKRhxQAyONNwW-Fho0R&export=download
Resolving drive.usercontent.google.com (drive.usercontent.google.com)... 2404:6800:4002:81b::2001, 172.217.26.33
Connecting to drive.usercontent.google.com (drive.usercontent.google.com)|2404:6800:4002:81b::2001|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 227054 (222K) [application/octet-stream]
Saving to: ‘marketing_data.csv’

marketing_data

In [24]:
df = pd.read_csv('./marketing_data.csv')
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   ID                   2240 non-null   int64
 1   Year_Birth           2240 non-null   int64
 2   Education            2240 non-null   str  
 3   Marital_Status       2240 non-null   str  
 4    Income              2216 non-null   str  
 5   Kidhome              2240 non-null   int64
 6   Teenhome             2240 non-null   int64
 7   Dt_Customer          2240 non-null   str  
 8   Recency              2240 non-null   int64
 9   MntWines             2240 non-null   int64
 10  MntFruits            2240 non-null   int64
 11  MntMeatProducts      2240 non-null   int64
 12  MntFishProducts      2240 non-null   int64
 13  MntSweetProducts     2240 non-null   int64
 14  MntGoldProds         2240 non-null   int64
 15  NumDealsPurchases    2240 non-null   int64
 16  NumWebPurchases      2240 non-null 

In [25]:
df.head()

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumStorePurchases,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Response,Complain,Country
0,1826,1970,Graduation,Divorced,"$84,835.00",0,0,6/16/14,0,189,...,6,1,0,0,0,0,0,1,0,SP
1,1,1961,Graduation,Single,"$57,091.00",0,0,6/15/14,0,464,...,7,5,0,0,0,0,1,1,0,CA
2,10476,1958,Graduation,Married,"$67,267.00",0,1,5/13/14,0,134,...,5,2,0,0,0,0,0,0,0,US
3,1386,1967,Graduation,Together,"$32,474.00",1,1,5/11/14,0,10,...,2,7,0,0,0,0,0,0,0,AUS
4,5371,1989,Graduation,Single,"$21,474.00",1,0,4/8/14,0,6,...,2,7,1,0,0,0,0,1,0,SP


In [26]:
df.shape

(2240, 28)

# Random Variables (RV)

## Motivation

In probability, experiments produce outcomes.

Examples:

### Coin Toss

Possible outcomes:

$$
S = \{H, T\}
$$

### Dice Roll

Possible outcomes:

$$
S = \{1,2,3,4,5,6\}
$$

Working directly with outcomes is often inconvenient.

Instead, we assign numerical values to outcomes.

This assignment is called a **Random Variable**.

---

# Definition of a Random Variable

A Random Variable (RV) is a function that maps outcomes of a random experiment to numerical values.

Mathematically,

$$
X : S \rightarrow \mathbb{R}
$$

where:

- $S$ = Sample Space
- $\mathbb{R}$ = Set of Real Numbers

---

## Intuition

Think of a Random Variable as:

> A numerical representation of outcomes from a random experiment.

It converts information into numbers so that mathematical operations can be performed.

---

# Example 1: Coin Toss

Sample Space:

$$
S = \{H,T\}
$$

Define a random variable:

$$
X =
\begin{cases}
1 & \text{if Head occurs} \\
0 & \text{if Tail occurs}
\end{cases}
$$

Then:

$$
P(X=1)=\frac12
$$

$$
P(X=0)=\frac12
$$

This is a Random Variable because it maps outcomes to numbers.

---

# Example 2: Dice Roll

Sample Space:

$$
S=\{1,2,3,4,5,6\}
$$

Define:

$$
X = \text{Number appearing on the die}
$$

Possible values:

$$
X \in \{1,2,3,4,5,6\}
$$

Each value has probability:

$$
P(X=x)=\frac16
$$

for

$$
x \in \{1,2,3,4,5,6\}
$$

---

# Random Variables in Machine Learning

In Machine Learning, every feature can be viewed as a random variable.

Examples:

| Feature | Random Variable Type |
|----------|----------|
| Income | Numerical |
| Age | Numerical |
| Height | Numerical |
| Country | Categorical |
| Education | Categorical |
| Gender | Categorical |

Example Dataset:

| Income | Country |
|----------|----------|
| 100000 | India |
| 85000 | USA |
| 120000 | Germany |

Here:

- Income is a Random Variable
- Country is a Random Variable

Both contain information about observations.

---

# Types of Random Variables

Random Variables are broadly classified into:

1. Discrete Random Variables
2. Continuous Random Variables

---

# 1. Discrete Random Variable

A Random Variable that can take only countable values.

Examples:

### Dice

$$
X \in \{1,2,3,4,5,6\}
$$

### Number of Children

$$
X \in \{0,1,2,3,\ldots\}
$$

### Number of Cars Owned

$$
X \in \{0,1,2,\ldots\}
$$

Characteristics:

- Countable values
- Finite or countably infinite
- Gaps between possible values

---

## Examples

### Coin Toss

$$
X=
\begin{cases}
1 & \text{Head}\\
0 & \text{Tail}
\end{cases}
$$

Discrete values:

$$
\{0,1\}
$$

---

### Dice Roll

$$
X \in \{1,2,3,4,5,6\}
$$

Discrete values.

---

# Ordinal Random Variables

Some discrete variables have a natural ordering.

Examples:

Education Level

```text
High School
    ↓
Undergraduate
    ↓
Postgraduate
    ↓
PhD
```

These categories are ordered.

Therefore:

> Education is an Ordinal Random Variable.

---

# 2. Continuous Random Variable

A Random Variable that can take infinitely many values within an interval.

Examples:

- Income
- Height
- Weight
- Temperature
- Time

---

## Example: Income

Income may take values such as:

$$
50000
$$

$$
50000.25
$$

$$
50000.257
$$

$$
50000.25781
$$

and infinitely many others.

Therefore Income is a Continuous Random Variable.

---

## Example: Height

A person's height may be:

$$
170.1
$$

$$
170.15
$$

$$
170.153
$$

$$
170.15382
$$

Thus height is continuous.

---

# Categorical Random Variables

Categorical variables contain labels instead of numerical quantities.

Examples:

```text
Country:
{
India,
USA,
Germany,
Canada
}
```

```text
Gender:
{
Male,
Female,
Other
}
```

Categories may be:

### Nominal

No order exists.

Example:

```text
Country
{
India,
USA,
Germany
}
```

No logical ordering.

---

### Ordinal

Order exists.

Example:

```text
Education
{
High School,
Undergraduate,
Postgraduate,
PhD
}
```

Logical ordering exists.

---

# Continuous vs Discrete

| Discrete RV | Continuous RV |
|-------------|---------------|
| Countable values | Infinite values |
| Dice outcome | Income |
| Number of children | Height |
| Number of defects | Weight |
| Coin toss outcome | Temperature |

---

# Important Machine Learning Perspective

In Machine Learning, features are usually categorized as:

### Numerical Features

Examples:

- Income
- Age
- Height
- Salary

These are usually continuous random variables.

---

### Categorical Features

Examples:

- Country
- Gender
- Education

These are categorical random variables.

---

# One-Line Summary

> A Random Variable is a numerical representation of outcomes from a random experiment, and it may be discrete, continuous, nominal, or ordinal.

In [34]:
r=df['MntGoldProds'].mean()
print(r)

44.021875


In [35]:
r=df['MntFruits'].mean()
print(r)

26.302232142857143


In [36]:
r=df['MntWines'].mean()
print(r)

303.9357142857143


In [38]:
r=df['MntMeatProducts'].mean()
print(r)

166.95


In [39]:
df['MntWines']

0       189
1       464
2       134
3        10
4         6
       ... 
2235    372
2236      5
2237    185
2238    267
2239    169
Name: MntWines, Length: 2240, dtype: int64

In [40]:
print("Gold:", df['MntGoldProds'].max(), df['MntGoldProds'].min())
print("Fruits:", df['MntFruits'].max(), df['MntFruits'].min())
print("Sweets:", df['MntSweetProducts'].max(), df['MntSweetProducts'].min())
print("Wine:", df['MntWines'].max(), df['MntWines'].min())
print("Meat:", df['MntMeatProducts'].max(), df['MntMeatProducts'].min())
print("Fish:", df['MntFishProducts'].max(), df['MntFishProducts'].min())

Gold: 362 0
Fruits: 199 0
Sweets: 263 0
Wine: 1493 0
Meat: 1725 0
Fish: 259 0


# Mean is Not Enough

The mean summarizes the center of the data but does not describe the spread.

Example:

MntMeatProducts

Min = 0
Max = 1725

A very large maximum value suggests possible outliers.

Since the mean is sensitive to outliers, it may not represent a typical observation.

Therefore, we also compute:

- Minimum
- Maximum
- Median

Median is more robust to outliers and often provides a better estimate of a typical observation when extreme values exist.

In [41]:
print("Gold:", df['MntGoldProds'].mean(), df['MntGoldProds'].median())
print("Fruits:", df['MntFruits'].mean(), df['MntFruits'].median())
print("Sweets:", df['MntSweetProducts'].mean(), df['MntSweetProducts'].median())
print("Wine:", df['MntWines'].mean(), df['MntWines'].median())
print("Meat:", df['MntMeatProducts'].mean(), df['MntMeatProducts'].median())
print("Fish:", df['MntFishProducts'].mean(), df['MntFishProducts'].median())

Gold: 44.021875 24.0
Fruits: 26.302232142857143 8.0
Sweets: 27.06294642857143 8.0
Wine: 303.9357142857143 173.5
Meat: 166.95 67.0
Fish: 37.52544642857143 12.0


# Mean vs Median

Both Mean and Median are measures of central tendency.

They try to answer:

> What is the "typical" or "central" value of the dataset?

---

# Mean

The mean is the arithmetic average of all observations.

$$
\text{Mean}
=
\frac{x_1+x_2+\cdots+x_n}{n}
$$

where:

- $x_1, x_2, ..., x_n$ are observations
- $n$ is the total number of observations

---

## Example

Data:

$$
\{10,11,9,8,12\}
$$

Mean:

$$
\frac{10+11+9+8+12}{5}
=
\frac{50}{5}
=
10
$$

Thus,

$$
\text{Mean}=10
$$

---

# Problem with Mean

Mean is heavily affected by extreme values (outliers).

Consider:

$$
\{10,11,9,8,12,1000\}
$$

Mean:

$$
\frac{10+11+9+8+12+1000}{6}
=
\frac{1050}{6}
=
175
$$

---

## Observation

The original values are all close to:

$$
10
$$

but the mean becomes:

$$
175
$$

because of one extreme value:

$$
1000
$$

This value is called an **outlier**.

---

# Outlier

An outlier is an observation that is significantly different from most other observations.

Example:

$$
\{10,11,9,8,12,1000\}
$$

Here,

$$
1000
$$

is an outlier.

---

# Limitation of Mean

Mean is sensitive to outliers.

A few extreme observations can drastically change the mean.

Therefore:

> Mean may not always represent the typical observation in the dataset.

---

# Median

Median is the middle value after sorting the observations.

Unlike the mean, the median depends only on position, not magnitude.

---

## Step 1: Sort the Data

Data:

$$
\{10,11,9,12,8,1000\}
$$

Sorted:

$$
\{8,9,10,11,12,1000\}
$$

---

## Even Number of Observations

If the number of observations is even:

$$
n=6
$$

Median is the average of the two middle values.

Middle values:

$$
10,11
$$

Median:

$$
\frac{10+11}{2}
=
10.5
$$

Thus,

$$
\text{Median}=10.5
$$

---

# Compare Mean and Median

For:

$$
\{10,11,9,8,12,1000\}
$$

Mean:

$$
175
$$

Median:

$$
10.5
$$

---

## Observation

Mean suggests the central value is:

$$
175
$$

which is misleading.

Median suggests:

$$
10.5
$$

which is much closer to most observations.

---

# Why Median is Better Here

The median ignores the magnitude of extreme values.

Changing:

$$
1000
\rightarrow
100000
$$

would drastically change the mean.

However, the median would remain:

$$
10.5
$$

---

# Robust Estimator

Median is called a **robust estimator**.

A robust estimator is not heavily influenced by outliers.

Therefore:

> Median is more reliable when extreme values are present.

---

# When to Use Mean

Use Mean when:

- Data is approximately symmetric
- Few or no outliers exist
- Every observation should contribute equally

Examples:

- Exam scores
- Heights
- Controlled scientific measurements

---

# When to Use Median

Use Median when:

- Data contains outliers
- Data is skewed
- Extreme values exist

Examples:

- Income
- House prices
- Customer spending
- Medical costs

---

# Example from Marketing Dataset

For some product categories:

| Feature | Mean | Median |
|----------|----------|----------|
| Gold Products | 44.02 | 24 |
| Fruits | 26.30 | 8 |
| Sweet Products | 27.06 | 8 |
| Wines | 303.94 | 173.5 |
| Meat Products | 166.95 | 67 |
| Fish Products | 37.53 | 12 |

Observation:

The mean is much larger than the median for many features.

This suggests:

- Presence of outliers
- Right-skewed distributions
- A small number of customers spend very large amounts

---

# ML Perspective

Before training any model:

1. Compute Mean
2. Compute Median
3. Check Min and Max
4. Look for Outliers

If Mean and Median differ significantly:

> The data is likely skewed or contains outliers.

---

# Quick Revision

### Mean

$$
\text{Mean}
=
\frac{\sum x_i}{n}
$$

- Uses all observations
- Sensitive to outliers

---

### Median

Middle observation after sorting.

- Robust to outliers
- Better for skewed data

---

### Important Interview Question

Which measure is more robust to outliers?

**Answer: Median**

Why?

Because it depends on position rather than the magnitude of extreme values.

---

# One-Line Summary

> Mean uses all values and is sensitive to outliers, whereas Median is a robust measure of central tendency that is much less affected by extreme values.